include/nvexec/detail/memory.cuh

```diff
@@ -110,4 +116,243 @@ namespace nvexec::STDEXEC_STREAM_DETAIL_NS {
+      return this == &other;
+    }
+  };
+
+  struct synchronized_pool_resource : std::pmr::memory_resource {
+    constexpr static size_t alignment = 256;
+
+    struct block_descriptor_t {
+      static constexpr unsigned int min_bin = 3;
+
+      void* ptr{};
+      unsigned int bin{};
+      size_t bytes{};
+
+      block_descriptor_t(size_t bytes)
+        : bin(cuda::std::bit_width(bytes))
+        , bytes(1ull << bin) {
+        if (bin < min_bin) {
+          bin = min_bin;
+          bytes = 1ull << bin;
+        }
+      }
+
+      block_descriptor_t(void* ptr)
+        : ptr(ptr) {
+      }
+    };
+
+    struct ptr_comparator_t {
+      bool operator()(const block_descriptor_t& a, const block_descriptor_t& b) const {
+        return a.ptr < b.ptr;
+      }
+    };
+
+    struct size_comparator_t {
+      bool operator()(const block_descriptor_t& a, const block_descriptor_t& b) const {
+        return a.bytes < b.bytes;
+      }
+    };
+
+    using cached_blocks_t = std::multiset<block_descriptor_t, size_comparator_t>;
+    using busy_blocks_t = std::multiset<block_descriptor_t, ptr_comparator_t>;
+
+    std::mutex mutex;
+
+    std::pmr::memory_resource* upstream;
+    cached_blocks_t cached_blocks;
+    busy_blocks_t live_blocks;
+
+    synchronized_pool_resource(std::pmr::memory_resource* upstream)
+      : upstream(upstream)
+      , cached_blocks(size_comparator_t{})
+      , live_blocks(ptr_comparator_t{}) {
+    }
+
+    void* do_allocate(size_t bytes, size_t /* alignment */) override {
+      std::lock_guard<std::mutex> lock(mutex);
+
+      block_descriptor_t search_key{bytes};
+      cached_blocks_t::iterator block_itr = cached_blocks.lower_bound(search_key);
+
+      while ((block_itr != cached_blocks.end()) && (block_itr->bin == search_key.bin)) {
+        search_key = *block_itr;
+        live_blocks.insert(search_key);
+        cached_blocks.erase(block_itr);
+        return search_key.ptr;
+      }
+
+      search_key.ptr = upstream->allocate(search_key.bytes, alignment);
+      live_blocks.insert(search_key);
+      return search_key.ptr;
+    }
+
+    void do_deallocate(void* ptr, size_t /* bytes */, size_t /* alignment */) override {
+      std::lock_guard<std::mutex> lock(mutex);
+
+      block_descriptor_t search_key{ptr};
+      busy_blocks_t::iterator block_itr = live_blocks.find(search_key);
+
+      if (block_itr != live_blocks.end()) {
+        search_key = *block_itr;
+        live_blocks.erase(block_itr);
+        cached_blocks.insert(search_key);
+      }
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+
+    ~synchronized_pool_resource() {
+      while (!cached_blocks.empty()) {
+        cached_blocks_t::iterator begin = cached_blocks.begin();
+        upstream->deallocate(begin->ptr, begin->bytes, alignment);
+        cached_blocks.erase(begin);
+      }
+    }
+  };
```

Comment:
I desperately want to optimize this -- slab allocation! free lists! -- but it doesn't make sense unless a profiler says it does, I suppose.

Reply:
I wanted to have something simple for now to remove the initial managed transaction as fast as possible. I think we can improve this later.